ConwaysGameOfLife.Tests/GameControllerShould.cs

```diff
@@ -9,64 +10,121 @@ public class GameControllerShould
+        [Theory]
+        [InlineData("3,3", "💀💀💀\n💀💀💀\n💀💀💀\n")]
+        [InlineData("3,2", "💀💀💀\n💀💀💀\n")]
+        [InlineData("9,9", "💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n💀💀💀💀💀💀💀💀💀\n")]
+        public void DisplayUniverse_GivenSuccessfulCreationOfInitialUniverse(string input, string expected) 
         {
-            var universe = new Universe(3, 3);
-            var controller = new GameController(universe, _output, _input);
-            var expected = "💀💀💀\n💀💀💀\n💀💀💀\n";
-
-            controller.DisplayUniverse();
-
-            Assert.Equal(expected, _output.GetWriteLine());
+            _input.GetReadLine(input);
+            _controller.CreateInitialUniverse();
+            
+            _controller.DisplayUniverse();
+            var actual = _output.GetLastWriteLine();
+            
+            Assert.Equal(expected, actual);
         }
 
         [Theory]
-        [InlineData("3,3", 3, 3)]
-        [InlineData("10,10", 10, 10)]
-        public void ReturnUniverseString_GivenValidInput(string input, int gridWidth, int gridLength)
+        [InlineData("3,3")]
+        [InlineData("10,10")]
+        public void ReturnUniverseString_GivenValidInput(string input)
         {
-            var universe = new Universe(gridWidth, gridLength);
-            var controller = new GameController(universe, _output, _input);
+            InputParser.ParseUniverse(input);
             _input.GetReadLine(input);
             var expected = input;
 
-            var result = controller.CreateValidUniverseString();
+            var result = _controller.CreateValidUniverseString();
 
             Assert.Equal(expected, result);
         }
 
-        //TODO: how to test error message in CreateValidUniverseString
+        [Theory]
+        [InlineData(new string[] { "0,3", "3,0" }, "1,1")]
+        [InlineData(new string[] { "0;3", "3-0", "4 4" }, "3,3")]
+        [InlineData(new string[] { ",," }, "10,10")]
+        [InlineData(new string[] { "3,3," }, "10,10")]
+        [InlineData(new string[] { "3,3,3", "3;3;" }, "3,3")]
+        [InlineData(new string[] { "a,b" }, "3,3")]
+        [InlineData(new string[] { "-1,4", "4,-1" }, "3,3")]
+        [InlineData(new string[] { "10,0", "0,10" }, "3,3")]
+        public void ReturnUniverseString_GivenInvalidInputsFollowedByValidInput(IEnumerable<string> invalidInputs, string validInput)
+        {
+            _input.GetReadLine(invalidInputs);
+            _input.GetReadLine(validInput);
+            var expectedOutputLine1 = "Invalid input. Please try again.";
+            var expectedOutputLine2 = "Please enter the width & length for the Universe as a number followed by a comma then a number (e.g. '0,0').";
+
+            var result = _controller.CreateValidUniverseString();
+            var actualOutputLine1 = _output.GetWriteLine(0);
+            var actualOutputLine2 = _output.GetWriteLine(1);

```

Comment:
It would be nice to use dynamic indices for these two so that it's testing against the output from the last invalid input. At the moment it's testing the output only the first invalid input.